### PR TITLE
transfer the apikey with ciphertext to the enroll app

### PR DESCRIPTION
**Pull request by bajo14**
- process apikey encryption/decryption
- add sample_rijndael128GCM_decrypt

test: passed the uinttest

Signed-off-by: bajo14 <z_sr@msn.com>

### CODE REVIEW

File: ehsm_kms_service/enroll_app/enroll_msg.cpp [hunk 1]
```diff
@@ -846,22 +843,49 @@ enroll_status_t ra_proc_msg3_get_att_result_msg(RetJsonObj retJsonObj_msg3, std:
...
+enroll_status_t ra_proc_apikey_result_msg_get_apikey(RetJsonObj retJsonObj_apikey_result_msg, uint8_t *apikey)
 {
+    // check nonce
     sgx_quote_nonce_t nonce;
+    enroll_status_t enroll_ret = ENL_OK;
+    uint8_t *iv = nullptr;
+    uint8_t *mac = nullptr;
+    uint32_t ret = 0;
+    uint8_t cipherapikey[EH_API_KEY_SIZE + SAMPLE_SP_IV_SIZE + SAMPLE_AESGCM_MAC_SIZE] = {0};
+
```
Comment (syan10) [inline ehsm_kms_service/enroll_app/enroll_msg.cpp:855]: check the input parameters

Comment (syan10) [reply ehsm_kms_service/enroll_app/enroll_msg.cpp:855]: you can add a open and do it in next patch

**Review by syan10 — APPROVED** [review]

File: ehsm_kms_service/enroll_app/enroll_msg.cpp [hunk 2]
```diff
@@ -846,22 +843,49 @@ enroll_status_t ra_proc_msg3_get_att_result_msg(RetJsonObj retJsonObj_msg3, std:
...
+                                        NULL, 0,
+                                        reinterpret_cast<uint8_t(*)[SAMPLE_AESGCM_MAC_SIZE]>(mac));
+    if (ret)
+    {
+        enroll_ret = ENL_ERROR_DECRYPT_APIKEY_FAILED;
+        goto OUT;
+    }
+OUT:
+    // clear the g_sp_db database after the attesation session finished.
+    memset(&g_sp_db, 0, sizeof(sp_db_item_t));
```
Comment (yang8621) [inline ehsm_kms_service/enroll_app/enroll_msg.cpp:889]: use explicit_bzero

File: core/App/ehsm_napi.cpp [hunk 3]
```diff
@@ -1380,13 +1389,16 @@ char *NAPI_RA_GET_API_KEY(const char *p_att_result_msg)
     retJsonObj.addData_uint8Array("nonce", pt_att_result_msg->platform_info_blob.nonce.rand, 16);
     retJsonObj.addData_string("appid", p_appid);
     retJsonObj.addData_string("apikey", (char*)p_apikey.data);
+    retJsonObj.addData_uint8Array("cipherapikey", cipherapikey.data, cipherapikey.datalen);
 
     log_d("apikey_result_msg: \n%s",retJsonObj.toChar());
     log_d("***NAPI_RA_GET_API_KEY end.");
 
 OUT:
+    explicit_bzero(cipherapikey.data, cipherapikey.datalen);
```
Comment (yang8621) [inline core/App/ehsm_napi.cpp:1398]: actually, no need to clear ciphertext

File: core/Enclave/enclave_hsm.cpp [hunk 4]
```diff
@@ -835,6 +839,29 @@ sgx_status_t enclave_generate_apikey(uint8_t *p_apikey, uint32_t apikey_len)
...
+    ret = sgx_rijndael128GCM_encrypt(&sk_key,
+                                     p_apikey, apikey_len,
+                                     cipherapikey,
+                                     iv, EH_AES_GCM_IV_SIZE,
+                                     NULL, 0,
+                                     reinterpret_cast<uint8_t (*)[EH_AES_GCM_MAC_SIZE]>(mac));
+    if (ret != SGX_SUCCESS) {
+        printf("error encrypting plain text\n");
+    }
+
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:864]: memset_s sk_key

File: core/App/ehsm_napi.cpp [hunk 5]
```diff
@@ -1380,13 +1389,16 @@ char *NAPI_RA_GET_API_KEY(const char *p_att_result_msg)
     retJsonObj.addData_uint8Array("nonce", pt_att_result_msg->platform_info_blob.nonce.rand, 16);
     retJsonObj.addData_string("appid", p_appid);
     retJsonObj.addData_string("apikey", (char*)p_apikey.data);
```
Comment (yang8621) [inline core/App/ehsm_napi.cpp:1391]: Do we still need to return the plaintext of apikey?

File: ehsm_kms_service/enroll_app/enroll_msg.cpp (hunk 1 above)
Comment (bajo14) [reply ehsm_kms_service/enroll_app/enroll_msg.cpp:855]: add check code

File: ehsm_kms_service/enroll_app/enroll_msg.cpp (hunk 2 above)
Comment (bajo14) [reply ehsm_kms_service/enroll_app/enroll_msg.cpp:889]: got it

File: core/App/ehsm_napi.cpp (hunk 3 above)
Comment (bajo14) [reply core/App/ehsm_napi.cpp:1398]: ok, delete it

File: core/Enclave/enclave_hsm.cpp (hunk 4 above)
Comment (bajo14) [reply core/Enclave/enclave_hsm.cpp:864]: add it

File: core/App/ehsm_napi.cpp (hunk 5 above)
Comment (bajo14) [reply core/App/ehsm_napi.cpp:1391]: nodejs  store it to db

Comment (yang8621) [reply core/App/ehsm_napi.cpp:1391]: Clear. Thanks.

**Review by yang8621 — APPROVED** [review]

File: ehsm_kms_service/enroll_app/main.cpp [hunk 6]
```diff
@@ -118,19 +119,24 @@ int main(int argc, char *argv[])
...
-    // TODO: decrypt APP ID and API Key
-    log_i("decrypt APP ID and API Key success.");
-    log_i("Third handle success.");
 
     printf("\nappid: %s\n", retJsonObj.readData_cstr("appid"));
-    printf("\napikey: %s\n\n", retJsonObj.readData_cstr("apikey"));
+    printf("\napikey: %s\n\n", apikey);
+
+    log_i("decrypt APP ID and API Key success.");
+    log_i("Third handle success.");
```
[the remark this answers is not in the document]
Comment (bajo14) [reply ehsm_kms_service/enroll_app/main.cpp:139]: will add free handle

File: core/App/ehsm_napi.cpp [hunk 7]
```diff
@@ -1380,13 +1389,15 @@ char *NAPI_RA_GET_API_KEY(const char *p_att_result_msg)
...
+    retJsonObj.addData_uint8Array("cipherapikey", cipherapikey.data, cipherapikey.datalen);
 
     log_d("apikey_result_msg: \n%s",retJsonObj.toChar());
     log_d("***NAPI_RA_GET_API_KEY end.");
 
 OUT:
     explicit_bzero(p_apikey.data, p_apikey.datalen);
     SAFE_FREE(pt_att_result_msg);
+    SAFE_FREE(cipherapikey.data);
     SAFE_FREE(p_apikey.data);
```
Comment (syan10) [inline core/App/ehsm_napi.cpp:1401]: does the duplicate for the line#1388 and line#1390?

Comment (bajo14) [reply core/App/ehsm_napi.cpp:1401]: seems it is

**Review by syan10 — APPROVED** [review]

**Pull request merged**